src/Sarif.Multitool/Rules/SARIF2004.OptimizeFileSize.cs

```diff
@@ -55,6 +71,66 @@ protected override void Analyze(Run run, string runPointer)
         {
             AnalyzeLocationOnlyArtifacts(run, runPointer);
             AnalyzeIdOnlyRules(run, runPointer);
+
+            if (run.Results != null)
+            {
+                AnalyzeResults(run.Results, runPointer.AtProperty(SarifPropertyName.Results));
+            }
+        }
+
+        private void AnalyzeResults(IList<Result> results, string resultsPointer)
+        {
+            for (int i = 0; i < results.Count; i++)
+            {
+                string resultPointer = resultsPointer.AtIndex(i);
+                Result result = results[i];
+                ReportUnnecessaryAnalysisTarget(result, resultPointer);
+                ReportRuleDuplication(result, resultPointer);
+            }
+        }
+
+        private void ReportRuleDuplication(Result result, string resultPointer)
+        {
+            if (result.Rule != null)
+            {
+                if (!string.IsNullOrWhiteSpace(result.RuleId) || result.RuleIndex >= 0)
+                {
+                    //  {0}: This result specifies both 'result.ruleId' and 'result.rule'.
+                    // Prefer 'result.ruleId' because it is shorter and just as clear.
+                    LogResult(
+                        resultPointer,
+                        nameof(RuleResources.SARIF2004_OptimizeFileSize_Warning_AvoidDuplicativeResultRuleInformation_Text));
+                }
+                else if (result.Rule.ToolComponent == null)
+                {
+                    // {0}: This result uses the 'rule' property to specify the rule
+                    // metadata, but the 'ruleId' property suffices because the rule
+                    // is defined by 'tool.driver'. Prefer 'result.ruleId' because it
+                    // is shorter and just as clear.
+                    LogResult(
+                        resultPointer,
+                        nameof(RuleResources.SARIF2004_OptimizeFileSize_Warning_PreferRuleId_Text));
+                }
+            }
+        }
+
+        private void ReportUnnecessaryAnalysisTarget(Result result, string resultPointer)
+        {
+            if (result.Locations != null && result.AnalysisTarget != null)
+            {
+                foreach (Location location in result.Locations)
+                {
+                    if (location?.PhysicalLocation?.ArtifactLocation?.Uri == result.AnalysisTarget.Uri)
+                    {
+                        // {0}: The 'analysisTarget' property '{1}' is unnecessary because it is the same as
+                        // the result location. Remove the 'analysisTarget' property.
+                        LogResult(
+                            resultPointer.AtProperty(SarifPropertyName.AnalysisTarget),
+                            nameof(RuleResources.SARIF2004_OptimizeFileSize_Warning_AvoidDuplicativeAnalysisTarget_Text),
+                            result.AnalysisTarget.Uri.OriginalString);
```

Comment:
>result [](start = 28, length = 6)

I would `break` after logging the first one. If you discover that "analysisTarget isn't necessary because it's the same as location 0", there's no need to also report "analysisTarget isn't necessary because it's the same as location 1".

Reply:
This was marked Resolved but it wasn't actually fixed. I think you tried to fix it with the `return` statements above, but that's not what I meant.

---
In reply to: [455252796](https://github.com/microsoft/sarif-sdk/pull/1995#discussion_r455252796) [](ancestors = 455252796)

Reply:
i see! i though it was in another place.

---
In reply to: [455344867](https://github.com/microsoft/sarif-sdk/pull/1995#discussion_r455344867) [](ancestors = 455344867,455252796)